File: src/ginger/hardware/pca9685.py
"""PCA9685 16-channel 12-bit PWM driver over I2C."""

import math
import time
import smbus2


_MODE1 = 0x00
_PRESCALE = 0xFE
_LED0_ON_L = 0x06

_OSC_FREQ = 25_000_000  # 25 MHz internal oscillator
_PWM_RESOLUTION = 4096  # 12-bit


class PCA9685:
    def __init__(self, address: int = 0x40, bus: int = 1):
        self._bus = smbus2.SMBus(bus)
        self._address = address
        self._write(_MODE1, 0x00)

    def set_pwm_freq(self, freq: float) -> None:
        prescale = round(_OSC_FREQ / (_PWM_RESOLUTION * freq)) - 1
        old_mode = self._read(_MODE1)
        self._write(_MODE1, (old_mode & 0x7F) | 0x10)  # sleep
        self._write(_PRESCALE, prescale)
        self._write(_MODE1, old_mode)
        time.sleep(0.005)
        self._write(_MODE1, old_mode | 0x80)  # restart

    def set_pwm(self, channel: int, on: int, off: int) -> None:
        base = _LED0_ON_L + 4 * channel
        self._bus.write_i2c_block_data(
            self._address, base,
            [on & 0xFF, on >> 8, off & 0xFF, off >> 8],
        )

    def set_duty(self, channel: int, duty: int) -> None:
        """duty in range [-4095, 4095]; 0 = brake (both high)."""
        if duty == 0:
            self.set_pwm(channel, 0, _PWM_RESOLUTION - 1)
        elif duty > 0:
            self.set_pwm(channel, 0, min(duty, _PWM_RESOLUTION - 1))
        else:
            self.set_pwm(channel, 0, min(-duty, _PWM_RESOLUTION - 1))

    def set_servo_pulse_us(self, channel: int, pulse_us: float) -> None:
        """pulse_us: pulse width in microseconds (500–2500 for standard servos at 50 Hz)."""
        off = int(pulse_us * _PWM_RESOLUTION / 20_000)
        self.set_pwm(channel, 0, off)

    def close(self) -> None:
        self._bus.close()

    def _write(self, reg: int, value: int) -> None:
        self._bus.write_byte_data(self._address, reg, value)

    def _read(self, reg: int) -> int:
        return self._bus.read_byte_data(self._address, reg)

File: src/hal/pca9685.rs
//! PCA9685 16-channel 12-bit PWM driver over I2C.

use std::thread::sleep;
use std::time::Duration;

use rppal::i2c::I2c;

use crate::Result;

const MODE1: u8 = 0x00;
const PRESCALE: u8 = 0xFE;
const LED0_ON_L: u8 = 0x06;

const OSC_FREQ: u32 = 25_000_000;
const PWM_RES: u32 = 4096;

pub const MAX_DUTY: u16 = (PWM_RES - 1) as u16; // 4095
pub const FULL_OFF: u16 = PWM_RES as u16; // 4096 — sets FULL-OFF bit in OFF_H register

pub struct Pca9685 {
    i2c: I2c,
}

impl Pca9685 {
    pub fn new(address: u16) -> Result<Self> {
        let mut i2c = I2c::new()?;
        i2c.set_slave_address(address)?;
        i2c.smbus_write_byte(MODE1, 0x00)?;
        Ok(Self { i2c })
    }

    pub fn set_pwm_freq(&mut self, freq: f32) -> Result<()> {
        let prescale =
            ((OSC_FREQ as f32 / (PWM_RES as f32 * freq)).round() as u8).saturating_sub(1);
        let old_mode = self.i2c.smbus_read_byte(MODE1)?;
        self.i2c.smbus_write_byte(MODE1, (old_mode & 0x7F) | 0x10)?; // sleep
        self.i2c.smbus_write_byte(PRESCALE, prescale)?;
        self.i2c.smbus_write_byte(MODE1, old_mode)?;
        sleep(Duration::from_millis(5));
        self.i2c.smbus_write_byte(MODE1, old_mode | 0x80)?; // restart
        Ok(())
    }

    pub fn set_pwm(&mut self, channel: u8, on: u16, off: u16) -> Result<()> {
        let base = LED0_ON_L + 4 * channel;
        self.i2c.smbus_write_byte(base, (on & 0xFF) as u8)?;
        self.i2c.smbus_write_byte(base + 1, (on >> 8) as u8)?;
        self.i2c.smbus_write_byte(base + 2, (off & 0xFF) as u8)?;
        self.i2c.smbus_write_byte(base + 3, (off >> 8) as u8)?;
        Ok(())
    }

    /// pulse_us: pulse width in microseconds (500–2500 for standard servos at 50 Hz).
    pub fn set_servo_pulse_us(&mut self, channel: u8, pulse_us: f32) -> Result<()> {
        let off = (pulse_us * PWM_RES as f32 / 20_000.0) as u16;
        self.set_pwm(channel, 0, off)
    }
}
